apis/python/src/tiledbsoma/_dataframe.py

```diff
@@ -401,6 +401,37 @@ def get_enumeration_values(
 
         return self._handle.get_enumeration_values(column_names)
 
+    def extend_enumeration_values(
+        self: DataFrame, values: dict[str, pa.Array], deduplicate: bool = False
+    ) -> None:
+        """Extend enumeration values for each column defined in `values`.
+        Raises ``ValueError`` if any of the the specified column names is not in
+        the schema, or if any is not of Arrow dictionary type. All the values
+        must provided must not already exist in the schema (see the output of
+        ``get_enumeration_values``), unless ``deduplicate=True`` is provided."""
+        self.verify_open_for_writing()
+
+        # These assertions could be done in C++. However, it's easier here
+        # to do the exception-type multiplexing, raising ValueError for one
+        # thing, TileDBSOMAError for another.
+        for column_name in values.keys():
+            # As with get_enumeration_values: we are trusting pyarrow to raise
+            # KeyError, and raise it with a sufficiently clear error message,
+            # when the column name is not present within the schema.
+            field = self.schema.field(column_name)
+            if not pa.types.is_dictionary(field.type):
+                raise KeyError(
+                    f"schema column name '{column_name}' is not of dictionary type"
+                )
+            if pa.types.is_dictionary(values[column_name].type):
+                raise ValueError(
+                    f"value column name '{column_name}' is of dictionary type: pass its dictionary array instead"
+                )
+                # TO DISCUSS IN CODE REVIEW: alternatively:
```

Comment:
IMHO, you have it right.  The `values` argument should be a simple array of values.  Constraints on that array of values:
* Same primitive type as the dictionary value type, ie,. `self.schema[column_name].type.value_type == values.type`
* no dups unless `deduplicate` is True
As I read the code, both of these are satisfied....

Reply:
Yes, they are satisfied by the code that isn't commented out.

I didn't think about this as I coded it, but: if someone did pass us a categorical: we could penalize them by throwing, requiring that the do a `.categories` on their input, or, we could take the `.categories` _for them_ and continue. I can really go either way. The non-commented-out code is find; just food for thought.